### PR TITLE
Additive indexes containing multiple allowed groups

### PR DESCRIPTION
**Pull request by erikap**
Allow search indexes defined in `eager_indexing_groups` to partially match the user's allowed groups. Indexes will be combined at search time to fully match the incoming allowed groups. If no combination can be found a single index matching the user's allowed groups will be created.

### CODE REVIEW

File: lib/mu_search/config_parser.rb [hunk 1]
```diff
@@ -28,7 +28,6 @@ def self.parse(path)
         { name: "batch_size", parser: :parse_integer },
         { name: "max_batches", parser: :parse_integer },
         { name: "persist_indexes", parser: :parse_boolean },
-        { name: "additive_indexes", parser: :parse_boolean },
```
Comment (nvdk) [inline lib/mu_search/config_parser.rb:31]: isn't it odd this is removed? this is still configurable right?

**Comment by nvdk** [comment]
Will need some time to test and wrap my head around the changes, from afar this looks like a breaking change for existing configurations with `additive_indexes: false`. It might not be though :). 

Aside: Perhaps taking things a bit far, but (assuming this is already breaking) could it make sense to move the eager_indexes config into the index definition so you could configure this per type?

File: lib/mu_search/index_manager.rb [hunk 2]
```diff
@@ -40,21 +39,13 @@ def fetch_indexes(type_name, allowed_groups, force_update: false)
...
-                @logger.debug("INDEX MGMT") { "Fetched index for type '#{type_name}' and allowed_groups #{allowed_groups}: #{index.name}" }
-              end
+            additive_indexes = find_indexes_for_groups type_name, allowed_groups
+            indexes_to_update += additive_indexes
+            @logger.debug("INDEX MGMT") do
+              index_names_s = additive_indexes.map(&:name).join(", ")
+              "Fetched #{additive_indexes.length} additive indexes for type '#{type_name}' and allowed_groups #{allowed_groups}: #{index_names_s}\nMatching allowed groups: #{additive_indexes.map(&:allowed_groups).to_json}"
             end
-          elsif @indexes[type_name] # fetch all indexes, regardless of access rights
+          elsif @indexes[type_name] # fetch all indexes, regardless of access rights. Only used in a stack without mu-authorization.
```
Comment (nvdk) [inline lib/mu_search/index_manager.rb:44]: I don't think this comment holds. Afaik it's used when using "admin" routes such as delete, where you can delete all indexes for a certain type

**Comment by erikap** [comment]
> Will need some time to test and wrap my head around the changes, from afar this looks like a breaking change for existing configurations with `additive_indexes: false`. It might not be though :).
> 
Not breaking normally. In the current implementation with `additive_indexes: false` the `eager_indexing_groups` are assumed to be configured such that they fully match the user's allowed groups. When switching to this PR's implementation, combinations of indexes will be constructed at query time, but indexes who's allowed groups are a strict 'subset' of the allowed groups of another index will be discarded from the combination. So in the end, only 1 index will remain in the combination, namely the one configured in the `eager_indexing_groups` that fully matches the allowed groups. 
TLDR; logic to determine the indexes to be searched has changed, but the indexes that will actually be searched are still the same 

Change might be breaking if the assumption that the `eager_indexing_groups` fully match the user groups doesn't hold, but that looks more like a misconfiguration in the app than a breaking change to me.

> Aside: Perhaps taking things a bit far, but (assuming this is already breaking) could it make sense to move the eager_indexes config into the index definition so you could configure this per type?

Would indeed by a nice extension and not so difficult to add anymore. Can be implemented as non-breaking change by using the root `eager_indexing_groups` as fallback if they are not configured on the type.

File: lib/mu_search/index_manager.rb (hunk 2 above)
Comment (erikap) [reply lib/mu_search/index_manager.rb:44]: You're right. We checked the search-endpoint, where we use `get_allowed_groups_with_fallback` to get the allowed groups (from the incoming request with a fallback to execute a dummy SPARQL query and getting the groups from that response).  In that case you will only end without allowed groups if mu-authorization isn't  part of the stack.
But for the 'admin' endpoints we use `get_allowed_groups` instead of  `get_allowed_groups_with_fallback`. Will remove the comment and document in the README that those admin endpoints should not be publicly exposed.
